### ISSUE TITLE
YOLOv4 custom training -> tf -> tflite -> list index out of range on detection

### ISSUE THREAD

**Issue by daohoangson** [hunglc007/tensorflow-yolov4-tflite#269]
*Hi there, I'm pretty new to computer vision so please bear with me.*

1. I'm following [this](https://blog.roboflow.com/training-yolov4-on-a-custom-dataset/) tutorial to train a custom YOLOv4 model. Everything went well, the model performs decently.
2. Then I tried using your repo to convert the weights into tf and the tf model works similarly to darknet.
3. Finally, I tried converting the tf model into tflite. The conversion itself went fine but detection always failed with this:

```bash
python detect.py --weights ./checkpoints/yolov4-416.tflite --size 416 --model yolov4 --framework tflite --image /path/to/image.jpg
```

```
Traceback (most recent call last):
  File "detect.py", line 95, in <module>
    app.run(main)
  File "/usr/local/lib/python3.6/dist-packages/absl/app.py", line 300, in run
    _run_main(main, args)
  File "/usr/local/lib/python3.6/dist-packages/absl/app.py", line 251, in _run_main
    sys.exit(main(argv))
  File "detect.py", line 65, in main
    boxes, pred_conf = filter_boxes(pred[0], pred[1], score_threshold=0.25, input_shape=tf.constant([input_size, input_size]))
IndexError: list index out of range
```

Searching the issues yielded nothing so I went ahead and edited the `detect.py` file, replacing this:

```py
        if FLAGS.model == 'yolov3' and FLAGS.tiny == True:
            boxes, pred_conf = filter_boxes(pred[1], pred[0], score_threshold=0.25, input_shape=tf.constant([input_size, input_size]))
        else:
            boxes, pred_conf = filter_boxes(pred[0], pred[1], score_threshold=0.25, input_shape=tf.constant([input_size, input_size]))
```

With this:

```py
        for value in pred:
             boxes = value[:, :, 0:4]
             pred_conf = value[:, :, 4:]
```

Basically I copied the box logic from tf mode into tflite mode. And it seemed to work! So what does this mean? Did I do something wrong... or right?

**Comment by daohoangson** [comment]
It turned out I was wrong. What I did:

```bash
python save_model.py --weights best.weights --output ./checkpoints/yolov4-tiny-416 --input_size 416 --model yolov4 --tiny
python convert_tflite.py --weights ./checkpoints/yolov4-tiny-416--output ./checkpoints/yolov4-tiny-416.tflite
```

Basically I converted darknet weights into TensorFlow then converted that tf model into tflite. Which is incorrect. This is how to do it:

```
python save_model.py --weights best.weights --output ./checkpoints/yolov4-tiny-416-tflite --input_size 416 --model yolov4 --tiny --framework tflite
python convert_tflite.py --weights ./checkpoints/yolov4-tiny-416-tflite --output ./checkpoints/yolov4-tiny-416.tflite
```

Notice the `--framework tflite` at the end of save model command. Yeah, the line is quite long and it was cutoff in the README so I probably missed it. Closing this issue now...

**Comment by roderick1014** [comment]
@daohoangson  
Hi, I'm wondering how to get the ".weights" file.  I didn't get this kind of file after training.

**Comment by stoxxy** [comment]
> @daohoangson Hi, I'm wondering how to get the ".weights" file. I didn't get this kind of file after training.

https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwjI1YK8paj3AhVAAxAIHSlBBKoQFnoECAkQAQ&url=https%3A%2F%2Fgithub.com%2FAlexeyAB%2Fdarknet%2Freleases%2Fdownload%2Fdarknet_yolo_v3_optimal%2Fyolov4.weights&usg=AOvVaw30if4joxtTaS8DAh12vYQ4
